MAINT: stats: avoid spurious warnings in ncx2.pdf

sethtroisi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Fixes https://github.com/scipy/scipy/issues/5202

#### What does this implement/fix?
<!--Please explain your changes.-->

#### Additional information
<!--Any additional information you think is important.-->

This is my first change to a stats distributions so please excuse any misunderstandings about log_pdf vs pdf priority.

sethtroisi commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -554,18 +555,20 @@ def _parse_args_stats(self, %(shape_arg_str)s %(locscale_in)s, moments='mv'):
-def _ncx2_pdf(x, df, nc):
-    return np.exp(_ncx2_log_pdf(x, df, nc))
+    corr = ive(df2, xs*ns) / 2.0
+    return np.exp(res) * corr
```
In the `pdf` case this is an improvement. 
old code: `np.exp(res + np.log(ive))`
new code: `np.exp(res) * ive`

for the `log_pdf` case this is a regression
old code: `res + np.log(ive)`
new code: `_lazywhere(np.log(np.exp(res) * ive))`

sethtroisi commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -554,18 +555,20 @@ def _parse_args_stats(self, %(shape_arg_str)s %(locscale_in)s, moments='mv'):
 # The function name ncx2 is an abbreviation for noncentral chi squared.
 
 def _ncx2_log_pdf(x, df, nc):
+    pdf = _ncx2_pdf(x, df, nc)
+    return _lazywhere(pdf > 0, (pdf,), f=np.log, fillvalue=-np.inf)
```
`_ncx2_log_pdf` is used needed for [`ncx2._logpdf`](https://github.com/scipy/scipy/blob/1bce83f54924be8844b1b0ad67e528607eaebd61/scipy/stats/_continuous_distns.py#L5764) which is still useful(?) because it allows for calling `chi2.logpdf` more directly in the `nc==0` case.

I suspect right now in the `nc==0` case
`ncx2.logpdf(x, df, 0)` => `chp.logpdf(x, df))`

If I delete this and `ncx2._log_pdf` I think what will happen is
`ncx2.logpdf(x, df, 0)` => `np.log(ncx2.pdf(x, df, 0))` => `np.log(chi.pdf(x, df))` =>  `np.log(np.exp(chp.logpdf(x, df)))`

Opinions?

rlucas7 commented:
Thanks for looking into this @sethtroisi from what I gather of your investigation so far it looks like the changes you looked into involve tradeoffs in accuracy of `pdf` vs `logpdf` methods. 
I suspect (but haven't confirmed) this is because of limits with the used approximation. 

I'll leave here some of my reasoning for others to comment if they disagree but from what I can see I'd prefer not to take the approach you're currently following. I'd rather look into different tail approximations to the `ncx2` which may be as (or more) accurate in the tails. Please let me know if you disagree. 

The caution here on numeric sensitivity of the tail is that `ncx2` calculations are ***heavily*** used in statistics for power curve calculations so it is reasonable to subject this distribution to a high degree of scrutiny, especially in the tails. 

From the (admittedly brief) investigation I did it looks like we use the Poisson mixutre of chi-squared distributions to approximate the (log)pdf of the non-central chi-squared distribution.

That call involves a call to `ive` in `scipy.special` which is the root cause of the warning. 
`ive` generates a `0` like so:

```
>>> from scipy.stats import ncx2
>>> ncx2.pdf(1, 311, 2)      
9.017e-321
>>> ncx2.pdf(1, 312, 2)      
5.14e-322
>>> ncx2.pdf(1, 313, 2)      
3e-323
>>> ncx2.pdf(1, 314, 2)      
0.0
>>> ncx2.pdf(1, 315, 2)      
0.0
>>> ncx2.pdf(1, 316, 2)      
0.0
>>> ncx2.pdf(1, 317, 2)      
0.0
>>> ncx2.pdf(1, 318, 2)      
/Users/rlucas/scipy-dev/scipy/scipy/stats/_distn_infrastructure.py:563: RuntimeWarning: divide by zero encountered in log
  res += np.log(ive(df2, xs*ns) / 2.0)
0.0
>>> 
``` 
where `df2` is calculated here:
https://github.com/scipy/scipy/blob/f4b5605031f738bc87ae4e193d614d525b98ffba/scipy/stats/_distn_infrastructure.py#L560

so `df=318` becomes `df2=158` which is then passed to `ive(158, 1.4142135623730951)` which will return `0`. So my guess is that `ive` doesn't work well in the tails. 
If I knew fortran better I'd investigate the functions that are called in `ive`, like this one:
https://github.com/scipy/scipy/blob/8dba340293fe20e62e173bdf2c10ae208286692f/scipy/special/amos_wrappers.c#L300

If we can preserve the tail behavior for large `df` in `ncx2` I'd prefer that perhaps by looking at a different approximation.

It looks like there have been several authors looking into this via saddlepoint approximations, Laguerre approximations, normal approximations, and others. If you're interested to investigate the feasibility of those methods I can forward you the papers I've found, otherwise I'll take a look when I get some time (maybe during this summer).  

Of course appropriate cutoffs for `df`, `x`, and `nc` parameters would need to be determined but given the large `df` value here it seems like that would be appropriate.

The papers I have in mind to investigate are:

1. https://www.tandfonline.com/doi/abs/10.1080/03610919808813480
2.https://www.sciencedirect.com/science/article/abs/pii/S0167947308005653

probably a couple others if these don't yield much.

sethtroisi commented:
Thanks for the detailed write up @rlucas7, I'll read and parse this more tomorrow, but I think it's unrelated to what I'm changing.

I'm trying to avoiding `np.log(0)` by converting `np.log(np.exp(res + np.log(ive)))` into `np.log(res) * ive`

I believe the values should be the same for old and new code.

I tested in [this branch](https://github.com/scipy/scipy/compare/master...sethtroisi:maint_ncx2_warning_test?expand=1) that `assert_allclose(method1, method2, rtol=4e-15, atol=0)` over several million pseudo random calls.

rlucas7 commented:
From what I understood the changes you’ve proposed involve accuracy trade offs between the pdf and the logpdf methods. 

Did I misunderstand that point? 

Sincerely, 

-Lucas Roberts 

> On Jun 2, 2020, at 5:08 AM, Seth Troisi <notifications@github.com> wrote:
> 
> ﻿
> Thanks for the detailed write up @rlucas7, I'll read and parse this more tomorrow, but I think it's unrelated to what I'm changing.
> 
> I'm trying to avoiding np.log(0) by converting np.log(np.exp(res + np.log(ive))) into np.log(res) * ive
> 
> I believe the values should be the same for old and new code.
> 
> I tested in this branch that assert_allclose(method1, method2, rtol=4e-15, atol=0) over several million pseudo random calls.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

sethtroisi commented:
@rlucas7 I believe you have misunderstood the point; This change is just moving around `np.exp` and `np.log` to prevent `np.log(0)`.

At a high level the code used to be

```
def pdf(x):
   res = ... calculation ...
   return np.exp(res + np.log(ive)) # Runtimewarning if ive == 0
```

and the new code is

```
def pdf(x):
   res = ... calculation ...
   return np.exp(res) * ive
```

There should be no change in output ([tested here](https://github.com/scipy/scipy/compare/master...sethtroisi:maint_ncx2_warning_test?expand=1)), up to the limits of floating points (e.g. `exp(a + b) != exp(a) * exp(b)`)

rlucas7 commented:
Ah ok. I must’ve misunderstood the point. Thanks for clarifying. 

Sincerely, 

-Lucas Roberts 

>> On Jun 2, 2020, at 5:46 PM, Seth Troisi <notifications@github.com> wrote:
> ﻿
> @rlucas7 I believe you have misunderstood the point; This change is just moving around np.exp and np.log to prevent np.log(0).
> 
> At a high level the code used to be
> 
> def pdf(x):
>    res = ... calculation ...
>    return np.exp(res + np.log(ive)) # Runtimewarning if ive == 0
> and the new code is
> 
> def pdf(x):
>    res = ... calculation ...
>    return np.exp(res) * ive
> There should be no change in output (tested here), up to the limits of floating points (e.g. exp(a + b) != exp(a) * exp(b))
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

WarrenWeckesser commented:
By making this change, `logpdf(x, df, nc)` is calculated by computing the log of the PDF.  Generally, we don't want to do that, because we lose precision for very small values of the PDF.  For example, in scipy 1.5.0,

```
In [68]: ncx2.logpdf(2000, 3, 12)
Out[68]: -853.9352051903619
```

With this PR, we get

```
In [37]: ncx2.logpdf(2000, 3, 12)
Out[37]: -inf
```

because the PDF for those parameter values underflows to 0.

Instead of the change proposed here, we could probably just wrap the line that triggers the spurious warning in an `errstate` context in which the division by 0 is ignored.  That is, in `_ncx2_log_pdf`, change

```
    res += np.log(ive(df2, xs*ns) / 2.0)
```
to
```
    with np.errstate(divide='ignore'):
        res += np.log(ive(df2, xs*ns) / 2.0)
```

And to avoid future changes that might break `logpdf`, the calculation of `ncx2.logpdf(2000, 3, 12)` that is shown above could be added as a unit test.  I've verified with `mpmath` that the result is accurate to machine precision, so it can be used as the known correct value.

sethtroisi commented:
Thanks @WarrenWeckesser for the detailed write up and examples.

I had been using `scipy.stats.ncx2.logpdf(1, 411, 12)` as my failure case (where `corr -> 0` as `df` gets large)
I failed to consider that `res` can also head to zero.

I went with a different solution (I find catching error state to be ugly) and duplicated the internals of `_ncx2_log_pdf` into `_ncx2_pdf`

sethtroisi commented:
@WarrenWeckesser I updated this code; I duplicate `_ncx2_log_pdf` and change the return line. I also added a unit test from your example so this won't be broken in the future. if someone tries to combine.

WarrenWeckesser commented:
@sethtroisi, thanks, looks good.  It would be nice to eliminate the little bit of duplicated code, but that's something we can revisit when we start implementing more of the `stats` internal functions as ufuncs.

A nice future enhancement would be a `logive` or `logiv` function, so we can expand the parameter ranges over which `ive(...)` doesn't overflow in the expression `log(ive(df2, xs*ns) / 2.0)`, but that is probably a mini-research project.

sethtroisi commented:
```
import scipy.stats
import scipy.special

for x, df, nc in [
    [100, 5, 0.1], [100, 5, 1], [100, 5, 10], [100, 5, 100],
    [300, 10, 1], [500, 10, 1],
    [200, 100, 1], [500, 100, 1],
    [200, 300, 1], [500, 300, 1],
    [1, 100, 1], [1, 200, 1],
    [1000, 10, 2], [2000, 10, 2], [9000, 10, 2],
]:
    ncx2_log = scipy.stats.ncx2.logpdf(x, df, nc)
    ive_z = (x * nc) ** 0.5
    ive = scipy.special.ive(df/2.0 - 1, ive_z)
    print(f"\tncx2.logpdf(x={x:4}, df={df:3}, nc={nc:3}) =  {ncx2_log:8.2f} "
          f"  ive(df/2={df/2:5.1f}, {ive_z:6.2f}) = {ive:9.2e}")
```

```
	ncx2.logpdf(x= 100, df=  5, nc=0.1) =    -44.27   ive(df/2=  2.5,   3.16) =  1.54e-01
	ncx2.logpdf(x= 100, df=  5, nc=  1) =    -39.91   ive(df/2=  2.5,  10.00) =  1.14e-01
	ncx2.logpdf(x= 100, df=  5, nc= 10) =    -25.02   ive(df/2=  2.5,  31.62) =  6.87e-02
	ncx2.logpdf(x= 100, df=  5, nc=100) =     -3.92   ive(df/2=  2.5, 100.00) =  3.95e-02
	ncx2.logpdf(x= 300, df= 10, nc=  1) =   -125.28   ive(df/2=  5.0,  17.32) =  6.01e-02
	ncx2.logpdf(x= 500, df= 10, nc=  1) =   -219.24   ive(df/2=  5.0,  22.36) =  5.89e-02
	ncx2.logpdf(x= 200, df=100, nc=  1) =    -19.12   ive(df/2= 50.0,  14.14) =  1.35e-27
	ncx2.logpdf(x= 500, df=100, nc=  1) =   -122.76   ive(df/2= 50.0,  22.36) =  8.70e-21
	ncx2.logpdf(x= 200, df=300, nc=  1) =    -14.70   ive(df/2=150.0,  14.14) = 9.89e-141
	ncx2.logpdf(x= 500, df=300, nc=  1) =    -27.67   ive(df/2=150.0,  22.36) = 1.94e-114
	ncx2.logpdf(x=   1, df=100, nc=  1) =   -180.22   ive(df/2= 50.0,   1.00) =  1.08e-78
	ncx2.logpdf(x=   1, df=200, nc=  1) =   -429.45   ive(df/2=100.0,   1.00) = 6.23e-187
	ncx2.logpdf(x=1000, df= 10, nc=  2) =   -447.54   ive(df/2=  5.0,  44.72) =  4.99e-02
	ncx2.logpdf(x=2000, df= 10, nc=  2) =   -927.75   ive(df/2=  5.0,  63.25) =  4.42e-02
	ncx2.logpdf(x=9000, df= 10, nc=  2) =  -4354.13   ive(df/2=  5.0, 134.16) =  3.25e-02
```

I did a little research, underflow is happening when [ive's first param](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ive.html) (v = order) is large.

Looking at our current code AMOS [zbesi](http://netlib.org/amos/zbesi.f) it doesn't appear log result is supported.

https://cran.r-project.org/web/packages/Bessel/Bessel.pdf seems to support such a return (via besselI.nuAsym) and is GPL. R's manual also includes these [function](https://www.imsbio.co.jp/RGM/R_rdfile?f=Bessel/man/besselIasym.Rd&d=R_CC)

ev-br commented:
https://github.com/scipy/scipy/blob/master/scipy/special/_hyp0f1.pxd#L51 has some implementation of asymptotic expansions of modified Bessel Iv
